### PR TITLE
[Serve] Added support for composing arbitrary DAGs

### PR DESCRIPTION
**Pull request by alindkhare**
1. Enqueue methods only scheduled when oid available
2. Added support for directly specifying wall clock time deadline (crucial for pipelines)

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: python/ray/experimental/serve/examples/echo_pipeline.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+serve.link("my_endpoint4", "echo:v4")
+
+# get the handle of the endpoints
+handle1 = serve.get_handle("my_endpoint1")
+handle2 = serve.get_handle("my_endpoint2")
+handle3 = serve.get_handle("my_endpoint3")
+handle4 = serve.get_handle("my_endpoint4")
+
+start = time.time()
+print("Start firing to the pipeline: {} s".format(time.time()))
```
Comment (simon-mo) [inline python/ray/experimental/serve/examples/echo_pipeline.py]: Can you add the picture back?

File: python/ray/experimental/serve/request_params.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
+import inspect
+
+
+class RequestInObject:
```
Comment (simon-mo) [inline python/ray/experimental/serve/request_params.py]: -> RequestMetadata

File: python/ray/experimental/serve/handle.py [hunk 3]
```diff
@@ -47,12 +48,19 @@ def remote(self, *args, **kwargs):
             except ValueError as e:
                 raise RayServeException(str(e))
```
Comment (simon-mo) [inline python/ray/experimental/serve/handle.py]: - Can we add `self.slo_ms = ...` to the constructor.
- And `handle.options(slo_ms=new_slo_ms).remote()`
- `handle.option -> new handle` with a different slo_ms

Default slo per handle: `serve.get_handle("Name", slo_ms=100)` -> `handle.remote(...)`
Update slo at runtime: `handle.options(slo_ms=...).remote()`

File: python/ray/experimental/serve/request_params.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+        self.service = service
+        self.request_context = request_context
+        self.request_slo_ms = request_slo_ms
+        self.is_wall_clock_time = is_wall_clock_time
+        if request_slo_ms is None:
+            self.is_wall_clock_time = False
+
+    @classmethod
+    def get_kwargs(cls):
```
Comment (simon-mo) [inline python/ray/experimental/serve/request_params.py]: remove

File: python/ray/experimental/serve/request_params.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        request_slo_ms(float): Expected time for the query to get
+            completed.
+        is_wall_clock_time(bool): if True, router won't add wall clock
+            time to `request_slo_ms`.
+    """
+
+    def __init__(self,
+                 service,
+                 request_context,
+                 request_slo_ms=None,
```
Comment (simon-mo) [inline python/ray/experimental/serve/request_params.py]:
```
relative_slo_ms = None,
absoluate_slo_ms = None
```
Specify one of them, if both are specified, we will Assert error.

File: python/ray/experimental/serve/queues.py [hunk 6]
```diff
@@ -179,15 +179,19 @@ def _serve_metric(self):
...
-                              request_context,
-                              request_slo_ms=None):
+    async def enqueue_request(self, request_in_object, *request_args,
+                              **request_kwargs):
+        service = request_in_object.service
         logger.debug("Received a request for service {}".format(service))
 
-        request_slo_ms = _adjust_latency_slo(request_slo_ms)
+        # check if the slo specified is directly the
+        # wall clock time
```
Comment (simon-mo) [inline python/ray/experimental/serve/queues.py]: Let's move this to instance method for RequestMetadata:
```
def get_wall_clock_time(self):
     # Compute the wall_clock_time if the user specified relative time
```

File: python/ray/experimental/serve/server.py [hunk 7]
```diff
@@ -150,13 +152,17 @@ def __init__(self):
                 await JSONResponse({"error": str(e)})(scope, receive, send)
                 return
 
+        # create objects necessary for enqueue
```
Comment (simon-mo) [inline python/ray/experimental/serve/server.py]: Mark this as TODO

File: python/ray/experimental/serve/server.py [hunk 8]
```diff
@@ -150,13 +152,17 @@ def __init__(self):
...
+        # enclosing http_body_bytes due to
+        # https://github.com/ray-project/ray/issues/6944
+        args = (scope, [http_body_bytes])
+        kwargs = dict()
```
Comment (simon-mo) [inline python/ray/experimental/serve/server.py]: we don't need to pass in kawrgs, so we need this variable

File: python/ray/experimental/serve/queues.py [hunk 9]
```diff
@@ -179,15 +179,19 @@ def _serve_metric(self):
...
+        if request_in_object.is_wall_clock_time:
+            request_slo_ms = request_in_object.request_slo_ms
+        else:
+            request_slo_ms = _adjust_latency_slo(
+                request_in_object.request_slo_ms)
+        request_context = request_in_object.request_context
         query = Query(request_args, request_kwargs, request_context,
```
Comment (simon-mo) [inline python/ray/experimental/serve/queues.py]: request_args not needed

File: python/ray/experimental/serve/utils.py [hunk 10]
```diff
@@ -16,7 +16,8 @@ def parse_request_item(request_item):
     if request_item.request_context == TaskContext.Web:
         is_web_context = True
         asgi_scope, body_bytes = request_item.request_args
-        flask_request = build_flask_request(asgi_scope, io.BytesIO(body_bytes))
+        flask_request = build_flask_request(asgi_scope,
+                                            io.BytesIO(body_bytes[0]))
```
Comment (simon-mo) [inline python/ray/experimental/serve/utils.py]: Can you copy the TODO here as well. so we can fix it in the future

File: python/ray/experimental/serve/handle.py [hunk 11]
```diff
@@ -26,33 +27,52 @@ class RayServeHandle:
...
 
-    def __init__(self, router_handle, endpoint_name):
+    def __init__(self,
+                 router_handle,
+                 endpoint_name,
+                 relative_slo_ms=None,
+                 absolute_slo_ms=None):
         self.router_handle = router_handle
         self.endpoint_name = endpoint_name
+        assert (relative_slo_ms is None
```
Comment (simon-mo) [inline python/ray/experimental/serve/handle.py]: What happen if both are None

File: python/ray/experimental/serve/api.py [hunk 12]
```diff
@@ -413,7 +413,7 @@ def split(endpoint_name, traffic_policy_dictionary):
 
 
 @_ensure_connected
-def get_handle(endpoint_name):
+def get_handle(endpoint_name, relative_slo_ms=None, absolute_slo_ms=None):
     """Retrieve RayServeHandle for service endpoint to invoke it from Python.
 
     Args:
```
Comment (simon-mo) [inline python/ray/experimental/serve/api.py]: Add docs for the two new params

File: python/ray/experimental/serve/examples/echo_pipeline.py [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+             \          /
+              \        /
+               \      /
+                \    /
+                 \  /
+                  \/
+            "my_endpoint4"
+"""
+
+# an endpoint is associated with an http URL.
```
Comment (simon-mo) [inline python/ray/experimental/serve/examples/echo_pipeline.py]: can you use `@serve.route` to simplify `create_endpoint`, `create_backend` and `link`?

File: python/ray/experimental/serve/examples/echo_slo_reverse.py [hunk 14]
```diff
@@ -50,7 +59,7 @@ def echo_v1(flask_request, response="hello from python!"):
     print("> [REMOTE] Pinging handle.remote(response='{}',slo_ms={})".format(
         response, slo_ms))
     # slo can be specified via remote function
-    f = handle.remote(response=response, slo_ms=slo_ms)
+    f = handle.options(relative_slo_ms=slo_ms).remote(response=response)
```
Comment (simon-mo) [inline python/ray/experimental/serve/examples/echo_slo_reverse.py]: This is testing the slo override right? if so can you add a comment above if not can you test that case?

**Review by simon-mo — APPROVED** [review]
Great work, few small things and it's good to merge

File: python/ray/experimental/serve/examples/echo_pipeline.py (hunk 13 above)
Comment (alindkhare) [reply python/ray/experimental/serve/examples/echo_pipeline.py]: Nice suggestion! 👍

**Pull request merged**